### PR TITLE
[SDCISA-13746] Increase resilience on resource exhaustion

### PR DESCRIPTION
**Pull request by hiddenalpha**
The issue currently ([2024-03-08](https://xkcd.com/1179/)) can be observed here:
> oc -n isa-houston-test exec -ti "$(oc -n isa-houston-test get pods|egrep ston-[0-9]|egrep -v deplo|cut -f1 -d' ')" -- sh -c 'cd /usr/local/vertx/logs && tail -n42 -F houston.log | grep ConnectionPoolTooBusyException'

### CODE REVIEW

File: gateleen-routing/src/main/java/org/swisspush/gateleen/routing/Forwarder.java [hunk 1]
```diff
@@ -239,31 +241,49 @@ private void handleRequest(final HttpServerRequest req, final Buffer bodyData, f
...
             public void handle(AsyncResult<HttpClientRequest> event) {
                 req.resume();
 
                 if (event.failed()) {
-                    log.warn("Problem to request {}: {}", targetUri, event.cause());
+                    Throwable ex = event.cause();
+                    long postponeMs = 1000;
+                    long remainingTimeMs = timeoutMs - postponeMs;
+                    if (remainingTimeMs > 0 && ex instanceof ConnectionPoolTooBusyException) {
+                        log.debug("EAGAIN: No connection avail now for {}", targetUri, ex);
```
Comment (Kusig) [inline gateleen-routing/src/main/java/org/swisspush/gateleen/routing/Forwarder.java:267]: Why EAGAIN, is this a typo?

File: gateleen-routing/src/main/java/org/swisspush/gateleen/routing/Forwarder.java [hunk 2]
```diff
@@ -239,31 +241,49 @@ private void handleRequest(final HttpServerRequest req, final Buffer bodyData, f
...
             @Override
             public void handle(AsyncResult<HttpClientRequest> event) {
                 req.resume();
 
                 if (event.failed()) {
-                    log.warn("Problem to request {}: {}", targetUri, event.cause());
+                    Throwable ex = event.cause();
+                    long postponeMs = 1000;
+                    long remainingTimeMs = timeoutMs - postponeMs;
+                    if (remainingTimeMs > 0 && ex instanceof ConnectionPoolTooBusyException) {
```
Comment (Kusig) [inline gateleen-routing/src/main/java/org/swisspush/gateleen/routing/Forwarder.java:266]: What about the sizing of the ConnectionPool itself, must it be tuned accordingly to reduce the probability of this occurence?

File: gateleen-routing/src/main/java/org/swisspush/gateleen/routing/Forwarder.java (hunk 1 above)
Comment (Kusig) [inline gateleen-routing/src/main/java/org/swisspush/gateleen/routing/Forwarder.java:267]: Would this be worth of any metric in order to monitor the ConnectionPool Failures over time?

File: gateleen-routing/src/main/java/org/swisspush/gateleen/routing/Forwarder.java (hunk 2 above)
Comment (hiddenalpha) [reply gateleen-routing/src/main/java/org/swisspush/gateleen/routing/Forwarder.java:266]: Feel free to open a PR for it :)

File: gateleen-routing/src/main/java/org/swisspush/gateleen/routing/Forwarder.java (hunk 1 above)
Comment (hiddenalpha) [reply gateleen-routing/src/main/java/org/swisspush/gateleen/routing/Forwarder.java]: If so, we should measure how loaded the pool is. Not how often we end up in this code branch.

File: gateleen-routing/src/main/java/org/swisspush/gateleen/routing/Forwarder.java [hunk 3]
```diff
@@ -278,7 +298,7 @@ public void handle(AsyncResult<HttpClientRequest> event) {
                 }
                 setProfileHeaders(log, profileHeaderMap, cReq);
 
-                authHeader.ifPresent(authHeaderValue -> cReq.headers().set(authHeaderValue.key(), authHeaderValue.value()));
+                if (authHeader != null) cReq.headers().set(authHeader.key(), authHeader.value());
```
Comment (mcweba) [inline gateleen-routing/src/main/java/org/swisspush/gateleen/routing/Forwarder.java:301]: Why going from _Optional<AuthHeader>_ to _AuthHeader_ where you have to use _null_ checks again. I don't see how this makes the code better or more readable

Comment (hiddenalpha) [reply gateleen-routing/src/main/java/org/swisspush/gateleen/routing/Forwarder.java]: My opinion clearly differs in this regard. I just changed it back.

(HINT: Feel free to ignore my opinion that follows here)
<opinion>

Optional is nothing but adding unnecessary bloat in this case (*Just comparing both variants side-by-side should be enough to show why*). [I'm not alone](https://homes.cs.washington.edu/~mernst/advice/nothing-is-better-than-optional.html) with this opinion BTW. Yet another counter argument comes from IDEA itslef:

image

</opinion>

**Pull request closed without merging**